src/include/sof/audio/audio_stream.h

```diff
@@ -43,8 +47,13 @@
  * consumption/production and update the buffer state by calling
  * audio_stream_consume()/audio_stream_produce() (just a single call following
  * series of reads/writes).
+ *
+ * Audio stream implements pipeline2.0 sink and source API
  */
 struct audio_stream {
+	struct source source_api;	/**< source API */
+	struct sink sink_api;		/**< sink API */
```

Comment:
this doesn't seem right to me. These structures should be opaque. And you include them and the "internal" header into a public header, that "everybody" includes.

Reply:
Not everybody, just implementations of the sink/source
and I don't see any way to encapsulate it more because
1) implementations must implement all function pointers 
2) in most cases struct sink / struct source will be a static member of the implementation (as it is in audio_stream now). To make it static - at least a size of the structure must be known. It is no way in C to encapsulate it effectively.

In C++ is simple - implementations inherit from sink/source and implement protected virtual functions, having no access to private members. Unfortunately it is C :(
All I could do is to put struct source/sink into .c file and export create_() that would malloc space for structures. I really don't want to do this for many reasons.

I change sink_api_internals.h to sink_api_implementation.h and add some comments about it